spec/app/domain/authentication/authn_k8s/application_identity_spec.rb

```diff
@@ -0,0 +1,380 @@
+          end
+
+          context "has only a pod constraint" do
+            let(:host_annotations) { [namespace_annotation, pod_annotation, container_name_annotation] }
+
+            it "does not raise an error" do
+              expect { subject }.not_to raise_error
+            end
+          end
+
+          context "has only a deployment constraint" do
+            let(:host_annotations) { [namespace_annotation, deployment_annotation, container_name_annotation] }
+
+            it "does not raise an error" do
+              expect { subject }.not_to raise_error
+            end
+          end
+
+          context "has only a stateful set constraint" do
+            let(:host_annotations) { [namespace_annotation, stateful_set_annotation, container_name_annotation] }
+
+            it "does not raise an error" do
+              expect { subject }.not_to raise_error
+            end
+          end
+
+          context "has only a deployment config constraint" do
+            let(:host_annotations) { [namespace_annotation, deployment_config_annotation, container_name_annotation] }
+
+            it "does not raise an error" do
+              expect { subject }.not_to raise_error
+            end
+          end
+
+          context "has a valid constraint combination" do
+            let(:host_annotations) { [namespace_annotation, service_account_annotation, deployment_config_annotation, container_name_annotation] }
+
+            it "does not raise an error" do
+              expect { subject }.not_to raise_error
+            end
+          end
+        end
+
+        context "with multiple annotations for container name" do
+          let(:host_annotations) { [namespace_annotation,
+                                    container_name_annotation,
+                                    container_name_annotation_service_id_prefix,
+                                    container_name_annotation_kubernetes_prefix] }
+
+          it "does not raise an error" do
+            expect { subject }.not_to raise_error
+          end
+
+          it "chooses the container name from the service-id annotation" do
+            expect(subject.container_name).to eq("ServiceIdContainerName")
+          end
+        end
+      end
+
+      context "with an invalid application identity" do
+
+        context "where namespace constraint doesn't exist" do
+          let(:host_annotations) { [service_account_annotation, container_name_annotation] }
+
+          it "raises an error" do
+            expect { subject }.to raise_error(::Errors::Authentication::AuthnK8s::MissingNamespaceConstraint)
+          end
+        end
+
+        context "with a non existing resource" do
+          context "in a global constraint" do
+            let(:host_annotations) { [namespace_annotation, invalid_annotation, container_name_annotation] }
+
+            it "raises an error" do
+              expect { subject }.to raise_error(::Errors::Authentication::AuthnK8s::ScopeNotSupported)
```

Comment:
- Add test of annotation name with authn-k8sInvalid
- Add test of annotation name with authn-k8sInvalid/blah
- Add test of annotation name with authn-k8sInvalid/pod

Reply:
not sure where to put these tests as in all the cases above these annotations will be ingnored as they are not interpreted as part of the `authn-k8s` annotations (as expected)

Reply:
ran this test manually to verify it is ok but i won't add a UT for this.

Reply:
Added a UT: https://github.com/cyberark/conjur/blob/host-annotations/spec/app/domain/authentication/authn_k8s/application_identity_spec.rb#L348